services/src/machine_learning/mod.rs

```diff
@@ -0,0 +1,107 @@
+pub(crate) use xgboost_training::XgboostTrainingParams;
+
+/// Represents a single feature of a machine learning training data set
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningFeature {
+    pub feature_name: Option<String>, // e.g. temperature, elevation etc.
+    pub feature_data: Vec<f32>,
+}
+
+impl MachineLearningFeature {
+    pub fn new(feature_name: Option<String>, feature_data: Vec<f32>) -> MachineLearningFeature {
+        MachineLearningFeature {
+            feature_name,
+            feature_data,
+        }
+    }
+}
+
+/// This enum represents the different aggregators that can be used to initialize the
+/// different algorithms for collecting the data used in ml training.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub enum MachineLearningAggregator {
+    Simple,
+    ReservoirSampling,
+}
+
+/// The purpose of this trait is to facilitate a generic aggregation process of raster data.
+pub trait Aggregatable {
+    type Data;
+
+    /// This method should realize the aggregation algorithm of the implementing struct.
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<Self::Data>>);
+
+    /// Once the aggregation process is finished, return the data for further usage.
+    fn finish(&mut self) -> Vec<Self::Data>;
+
+    /// Used to generate a new instance of an aggregatable within a generic context
+    fn new(data: Vec<f32>, memory_limit: Option<usize>) -> Self;
+}
+
+/// A simple aggregator that just collects all incoming data in a vector.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SimpleAggregator {
+    pub data: Vec<f32>,
+}
+
+impl Aggregatable for SimpleAggregator {
+    type Data = f32;
+
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<f32>>) {
+        let v = incoming_data.into_iter().flatten().collect::<Vec<f32>>();
+
+        self.data.extend(v);
+    }
+
+    fn finish(&mut self) -> Vec<f32> {
+        std::mem::take(&mut self.data)
+    }
+
+    // TODO: implement memory limit
+    fn new(data: Vec<f32>, _memory_limit: Option<usize>) -> Self {
+        SimpleAggregator { data }
+    }
+}
+
+/// A reservoir sampling aggregator that samples a given number of elements from the incoming data.
+/// This aggregator can be used, when the data size exceeds the available memory.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct ReservoirSamplingAggregator {
+    pub data: Vec<f32>,
+}
+
+// TODO: implement reservoir sampling
+impl Aggregatable for ReservoirSamplingAggregator {
+    type Data = f32;
+
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<f32>>) {
+        let v = incoming_data.into_iter().flatten().collect::<Vec<f32>>();
+        self.data.extend(v);
+    }
+
+    fn finish(&mut self) -> Vec<f32> {
+        std::mem::take(&mut self.data)
+    }
```

Comment:
You should be able to do the following, if you adjust the trait accordingly
```suggestion
    fn finish(self) -> Vec<f32> {
        self.data
    }
```

Reply:
If I change the code like this, I'm getting this error:

```
 1  error[E0507]: cannot move out of `*agg` which is behind a mutable reference
    --> services/src/machine_learning/data_preparation.rs:120:34
     |
 120 |             let collected_data = agg.finish();
     |                                  ^^^^--------
     |                                  |   |
     |                                  |   `*agg` moved due to this method call
     |                                  move occurs because `*agg` has type `A`, which does not imple
 ment the `Copy` trait
     |
 note: `Aggregatable::finish` takes ownership of the receiver `self`, which moves `*agg`
    --> services/src/machine_learning/mod.rs:54:15
     |
 54  |     fn finish(self) -> Vec<Self::Data>;
```

Reply:
If you change `aggregator_vec.iter_mut()` to `aggregator_vec.into_iter()` you should be able to avoid this problem, because you then take ownership of the aggregator.